trinity/protocol/common/trackers.py

```diff
@@ -76,39 +80,43 @@ def get_stats(self) -> str:
         """
         if not self.total_msgs:
             return 'None'
-        avg_rtt = self.total_response_time / self.total_msgs
-        if not self.total_response_time:
-            items_per_second = 0.0
-        else:
-            items_per_second = self.total_items / self.total_response_time
+
+        try:
+            rt99 = self.round_trip_99th.value
+        except ValueError:
+            rt99 = 0
+
+        try:
+            rt_stddev = self.round_trip_stddev.value
+        except ValueError:
+            rt_stddev = 0
 
         # msgs: total number of messages
         # items: total number of items
-        # rtt: round-trip-time (avg/ema)
-        # ips: items-per-second (avg/ema)
+        # rtt: round-trip-time (ema/99th/stddev)
+        # ips: items-per-second (ema)
         # timeouts: total number of timeouts
         # missing: total number of missing response items
         # quality: 0-100 for how complete responses are
         return (
-            'msgs=%d  items=%d  rtt=%.2f/%.2f  ips=%.5f/%.5f  '
+            'msgs=%d  items=%d  rtt=%.2f/%.2f/%.2f  ips=%.5f  '
             'timeouts=%d  quality=%d'
         ) % (
             self.total_msgs,
             self.total_items,
-            avg_rtt,
             self.round_trip_ema.value,
-            items_per_second,
+            rt99,
+            rt_stddev,
             self.items_per_second_ema.value,
             self.total_timeouts,
             int(self.response_quality_ema.value),
         )
 
-    def record_timeout(self, timeout: float) -> None:
+    def record_timeout(self) -> None:
         self.total_msgs += 1
         self.total_timeouts += 1
         self.response_quality_ema.update(0)
         self.items_per_second_ema.update(0)
-        self.round_trip_ema.update(timeout)
```

Comment:
> More aggressively: we could reset the 99th percentile entirely on every timeout?

Going to take some time to think on this because I think the concerns are valid.  I'm :-1: on including timeouts in `rtt` metrics because they *polute* the data and our quality and timeout metrics should be capable of providing us with the data we need to make the adjustment.

Rough sketch of the idea that is forming.  New metric which is EMA of timeout percentage (like quality metric) and another new metric which is data for the last `M` requests on whether they timeout or not.  Then if we see the rate of timeouts in the last `M` requests grow beyond the EMA of percentages of requests that timeout by some amount we do something like resetting percentile stats under the assumption that networking conditions changed.

Reply:
> I'm  on including timeouts in `rtt` metrics because they _polute_ the data

Ah, actually, I was wrong about the syncing anyway, so no objections to removing `round_trip_ema` here. (Sync uses `items_per_second_ema`)

> Rough sketch of the idea that is forming

I think I see where you're going. We can maybe simplify it even more.

The tracker will currently adjust fairly quickly as long as at least *some* requests make it through. It's mostly a problem if we go to 100% timeouts. Even a timeout streak counter might work: if the last N(=5?) requests all time out, then reset.

Reply:
> Even a timeout streak counter might work: if the last N(=5?) requests all time out, then reset.

This is where I started but I a took it towards the `M/N` threshold idea because my intuition says 1) it isn't much more complex and 2) in the event that *some* requests are getting through it will still hit the threshold and trigger a reset which feels desirable to the converse of not resetting even though 80% of requests are timing out, just never in a full streak of 5 in a row.  I don't feel strongly but I do feel that either way implementing something like

```python
if timed_out:
    tracker.timeout_frequency_ema.update(1)
else:
    tracker.timeout_frequency_ema.update(0)
```

This gives us a much better/actionable metric than the current `total_timeouts` which feels kind of useless.

Reply:
I agree with the conclusion (the timeout frequency EMA is better than the timeout count).

> 2) in the event that _some_ requests are getting through it will still hit the threshold and trigger a reset which feels desirable to the converse of not resetting even though 80% of requests are timing out, just never in a full streak of 5 in a row.

Yeah, I think that's right. In the specific wifi scenario, I don't think it would matter, because even one or two requests that get through would be really slow, so it would bump the 99th high, quickly. But you could imagine other scenarios where a few get through fast, and the others time out.